### ISSUE TITLE
[yul] interpreter should not leave trace of side-effects of code that will be removed by the optimizer

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#6786]
## Description

(Found by ossfuzz nightly CI)

```
{
let a_0, _1, a
function foo_0(x_0,_2, x_2) -> x_6
{
let x_4 := extcodehash(extcodesize(1))
}
a_0 := foo_0(calldataload(0),0, a_0)
}
```

is optimized to

```
{}
```

I wonder why this was not a problem earlier. Wouldn't the redundant assignment eliminator have found this earlier?

## Environment

- Compiler version: latest develop

**Comment by bshastry** [comment]
This could just be because `a_0` is unused and the optimizer removes it altogether. This optimization was not happening before, so it is still a bug in the fuzzing setup.

**Comment by bshastry** [comment]
Trace of unoptimized code

```
Trace:
  EXTCODESIZE(1)
  EXTCODEHASH(16387318)
```

Trace of optimized code is empty of course.

**Comment by chriseth** [comment]
This is caused by #6746. `extcodehash` and `extcodesize` can be safely removed if their result is not used, so I think the optimizer is correct, but we probably need to adjust the trace so that it does not include these calls.

**Comment by bshastry** [comment]
Okay! For completeness, this happens for any instruction that has a side-effect but I think this should not matter as long as we remove unused side-effects from the trace.

**Comment by bshastry** [comment]
Should we add semantic tests for calls to the interpreter post optimization? Might help resolve issues like this in the future at the time of PR?

**Comment by chriseth** [comment]
There are tests for the interpreter, but probably not enough...

**Comment by chriseth** [comment]
Ah sorry, such a test would not have helped...

**Comment by bshastry** [comment]
Closed via #6798
